### PR TITLE
PE-421 Add enrollment report

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description:

This PR adds the Salesforce enrollment report backend.

## Notes:

New settings were added, check the README.md

## Reviewers:

- [ ] @andrey-canon 
- [ ] @diegomillan

### CODE REVIEW

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            user_salesforce_data = {
+                'FirstName': first_name if first_name else username,
+                'LastName': last_name if last_name else username,
+                'Email': user.get('email', ''),
+                'Company': self.extra_data.get('COMPANY_NAME', ''),
+                'Institution_Hidden': self.extra_data.get('INSTITUTION_HIDDEN_PREFIX', ''),
+                'Type_Hidden': self.extra_data.get('TYPE_HIDDEN', ''),
+                'Program_of_Interest': self.extra_data.get('PROGRAM_OF_INTEREST', ''),
+                'Intake_of_Intent': user.get('intake_of_intent', ''),
+                'Lead_Source': DEFAULT_LEAD_SOURCE,
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]:
```suggestion
                'Lead_Source': self.extra_data.get('LEAD_SOURCE', DEFAULT_LEAD_SOURCE)
```
and add lead_source to the readme file

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                'Secondary_Source': '',
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: we 'd better take this from the config file

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+                'Tertiary_Source': '',
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: same comment as above

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        Creates the enrollment list into Salesforce.
+
+        Args:
+            enrollment_data: List of list with the enrollment data.
+        Returns:
+            salesforce_result: List wiht the contact id data from Salesforce.
+        """
+        request_session = requests.Session()
+        salesforce_details = self._get_salesforce_details()
+        api_url = '{}{}'.format(
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py:154]: Add validation for self.extra_data.get('SALESFORCE', {}).get('SALESFORCE_API_ENROLLMENT_URL', '')

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            exit()
+
+        salesforce_result = []
+
+        for enrollments in enrollment_data:
+            request_data = json.dumps({
+                'enrollments': enrollments,
+            })
+
+            api_response = request_session.post(api_url, data=request_data)
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]:
```suggestion
            api_response = request_session.post(api_url, json=request_data)
```

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        print('Creating enrollments on Salesforce...')
+
+        if not api_url:
+            print('SALESFORCE_API_ENROLLMENT_URL was not provided.')
+            exit()
+
+        salesforce_result = []
+
+        for enrollments in enrollment_data:
+            request_data = json.dumps({
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: json.dumps is not necessary

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            salesforce_details.get('instance_url', ''),
+            self.extra_data.get('SALESFORCE', {}).get('SALESFORCE_API_ENROLLMENT_URL', ''),
+        )
+
+        request_session.headers.update({
+            'Authorization': '{} {}'.format(
+                salesforce_details.get('token_type', ''),
+                salesforce_details.get('access_token', ''),
+            ),
+            'Content-Type': 'application/json',
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: not necessary

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        if not api_url:
+            print('CONTACT_ID_API_URL was not provided.')
+            exit()
+
+        for data in users_data:
+            request_data = json.dumps({
+                'records': process_user_data(data),
+            })
+
+            response = request_session.post(api_url, data=request_data)
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: same comment as above

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+            'Authorization': 'Bearer {}'.format(self.settings.get('OPEN_EDX_OAUTH_TOKEN', '')),
+            'Content-Type': 'application/json',
+        })
+
+        if not api_url:
+            print('CONTACT_ID_API_URL was not provided.')
+            exit()
+
+        for data in users_data:
+            request_data = json.dumps({
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: same comment as above

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+        api_url = '{}{}'.format(
+            self.settings.get('LMS_URL', ''),
+            self.extra_data.get('CONTACT_ID_API_URL', '')
+        )
+        request_session.headers.update({
+            'Authorization': 'Bearer {}'.format(self.settings.get('OPEN_EDX_OAUTH_TOKEN', '')),
+            'Content-Type': 'application/json',
+        })
+
+        if not api_url:
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py]: check validation

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 1 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addresed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 2 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Actually they are not necessary.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 3 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Actually they are not necessary.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 4 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py:154]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 5 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 6 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 7 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 8 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 9 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 10 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py]: Addressed.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 2 above)
Comment (andrey-canon) [reply proversity_reports_script/report_backend/enrollment_report.py]: so remove it

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+            'client_id': salesforce_data.get('CLIENT_ID', ''),
+            'username': salesforce_data.get('USERNAME', ''),
+            'client_secret': salesforce_data.get('CLIENT_SECRET', ''),
+            'password': '{}{}'.format(
+                salesforce_data.get('PASSWORD', ''),
+                salesforce_data.get('SECURITY_TOKEN', ''),
+            ),
+        }
+
+        request_response = requests.post(salesforce_url, data=request_data)
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py:221]:
```suggestion
        request_response = requests.post(salesforce_url, json=request_data)
```

@Squirrel18

File: proversity_reports_script/report_backend/enrollment_report.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        """
+        print('Getting Salesforce token deatils...')
+
+        salesforce_data = self.extra_data.get('SALESFORCE', {})
+
+        if not salesforce_data:
+            print('SALESFORCE data was not provided.')
+            exit()
+
+        salesforce_url = salesforce_data.get('AUTHENTICATION_URL', '')
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/enrollment_report.py:204]: we should validate this data before use it  @Squirrel18

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 11 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py:221]: It doesn't need to be a json.

File: proversity_reports_script/report_backend/enrollment_report.py (hunk 12 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/enrollment_report.py:204]: Addressed.

**Review by Squirrel18 — APPROVED** [review]

**Pull request merged**